core/ipc/sink/common/src/main/java/org/opennms/core/ipc/sink/common/AsyncDispatcherImpl.java

```diff
@@ -67,226 +65,162 @@
     private final SyncDispatcher<S> syncDispatcher;
-    private OffHeapAdapter offHeapAdapter;
-    private ExecutorService offHeapAdapterExecutor = Executors.newSingleThreadExecutor();
     private final AsyncPolicy asyncPolicy;
-    private OffHeapQueue offHeapQueue;
-    private SinkModule<S,T> sinkModule;
-    private DispatcherState<W,S,T> state;
-    private boolean useOffHeap = false;
+    private final Counter droppedCounter;
     
-    final RateLimitedLog rateLimittedLogger = RateLimitedLog
+    private final DispatchQueue<S> dispatchQueue;
+    private final Map<String, CompletableFuture<DispatchStatus>> futureMap = new ConcurrentHashMap<>();
+    private final AtomicInteger activeDispatchers = new AtomicInteger(0);
+    
+    private final RateLimitedLog RATE_LIMITED_LOGGER = RateLimitedLog
             .withRateLimit(LOG)
             .maxRate(5).every(Duration.standardSeconds(30))
             .build();
 
-    final LinkedBlockingQueue<Runnable> queue;
-    final ExecutorService executor;
+    private final ExecutorService executor;
 
     public AsyncDispatcherImpl(DispatcherState<W, S, T> state, AsyncPolicy asyncPolicy,
-            SyncDispatcher<S> syncDispatcher) {
+                               SyncDispatcher<S> syncDispatcher) {
         Objects.requireNonNull(state);
         Objects.requireNonNull(asyncPolicy);
-        this.syncDispatcher = Objects.requireNonNull(syncDispatcher);
+        Objects.requireNonNull(syncDispatcher);
+        this.syncDispatcher = syncDispatcher;
         this.asyncPolicy = asyncPolicy;
-        this.state = state;
-        sinkModule = state.getModule();
-        if (OffHeapServiceLoader.isOffHeapEnabled()) {
-            offHeapQueue = OffHeapServiceLoader.getOffHeapQueue();
-            if (offHeapQueue != null) {
-                useOffHeap = true;
-                LOG.info("Offheap storage enabled for sink module, {}", sinkModule.getId());
-            }
-        }
-        
-        final RejectedExecutionHandler rejectedExecutionHandler;
-        if (asyncPolicy.isBlockWhenFull()) {
-            // This queue ensures that calling thread is blocked when the queue is full
-            // See the implementation of OfferBlockingQueue for details
-            queue = new OfferBlockingQueue<>(asyncPolicy.getQueueSize());
-            rejectedExecutionHandler = new ThreadPoolExecutor.AbortPolicy();
+        SinkModule<S, T> sinkModule = state.getModule();
+        Optional<DispatchQueueFactory> factory = DispatchQueueServiceLoader.getDispatchQueueFactory();
+
+        if (factory.isPresent()) {
+            LOG.debug("Using queue from factory");
+            dispatchQueue = factory.get().getQueue(asyncPolicy, sinkModule.getId(),
+                    sinkModule::marshalSingleMessage, sinkModule::unmarshalSingleMessage);
         } else {
-            queue = new LinkedBlockingQueue<Runnable>(asyncPolicy.getQueueSize());
```

Comment:
Consider using `asyncPolicy.getQueueSize()` as the default inMemoryQueueSize for the specific module. We can extend AsyncPolicy to set `offHeapSize` then everything is associated with module

Reply:
It should be easy to re-use the queue size value but what about the batch size and max file size? Do they belong in the XML configuration for the module? They are implementation specific details for this queueing strategy. Do you think it makes sense to move those into the per-module XML as well?

Another thing to think about... the original meaning of the queue size setting is not really applicable in the case of off-heap storage. For instance if the queue gets filled up it doesn't mean we are going to start rejecting or blocking, we will just be going off-heap. This was the original reasoning why I didn't re-use the existing config.

Reply:
Yeah, I was thinking that it makes sense to have module specific configuration for off-heap as well so that we can change off-heap size based on needs of module.  

But looking at configuration more closely, it may be confusing to add `batchSize` and `queueSize` in module configuration as they are also used for `aggregation`.  Let's keep them separate for now.

Reply:
Ok thanks for the feedback. I think ultimately what we would want is an aggregate max file size instead of a per-module file size so the user could choose to use a max of XGB across all modules. To keep it simple I just went with the fixed file size  per module for now.